### ISSUE TITLE
Trim analyzer: Doesn't produce warnings due to annotated generic parameter in various situations

### ISSUE THREAD

**Issue by vitek-karas** [dotnet/runtime#95121]
```csharp
Method<string>();

Type Method<TUnknown>()
{
    // Should WARN - TUnknown does have All annotation
    return typeof(TRequires<TUnknown>);
}

class TRequires<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] T>
{
}
```

The able produces IL2091 in the trimmer, but doesn't produce it in the analyzer. The bug seems to be that we don't process generic parameter annotations in a `typeof`.

Also happens for a declaration of a local variable:
```csharp
void Method<TUnknown>()
{
    // Trimmer produces IL2091 here, analyzer doesn't
    TRequires<TUnknown> a;
}
```

Note that for local variables NativeAOT doesn't produce the warning either - see the comment here for why:
https://github.com/dotnet/runtime/blob/7f818485a669eef2271034cb1aee3b01bdeeaef6/src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/GenericParameterWarningLocation.cs#L12-L28

We should figure out if analyzer needs to produce the warning or not.

Another situation is a static field access:
```csharp
    TRequires<TUnknown>.Field;  // Should warn
```

Yet another situation - expression trees:
```csharp
Expression<Func<string>> a = () => TRequires<TUnknown>.Field
```

And then type checks (NativeAOT doesn't warn here, see the link above):
```csharp
if (value is TRequires<TUnknown>) {}

var a = value as TRequires<TUnknown>;
```

Catch and filter expressions (NativeAOT doesn't warn):
```csharp
try {}
catch (TRequires<TUnknown>) {}
catch (Exception ex) when (ex is TRequires<TUnknown>) {}
```

Nested generic for methods:
```csharp
GenericMethod<Generic1<Generic2<TRequires<TUnknown>>>();
```

Similarly nested generics don't cause DAM marking and thus diagnostics:
```csharp
GenericMethod<Generic1<Generic2<TRequires<TypeWIthRUCMethod>>>(); // Should produce IL2026
```

**Comment by vitek-karas** [comment]
Good point - If we can get the same behavior as NativeAOT I think that would be great (it warns less). Also if we can clearly define the way to do that in the type system (NativeAOT implementation kind of fell out of the way the compiler works), we could apply it to trimmer.

I don't know the typeof behavior right now - the thing about typeof is that it's considered a "reflection access" in NativeAOT - so maybe that's the reason. I would have to dig into that. It could be an implementation detail. In general I agree that I don't see why it would need to warn - having the type accessible doesn't run code on it (yet). Accessing statics of instantiating it would do that - but we should see all of that when it's done via reflection anyway. But it's possible this was one of the "this just doesn't work well for real world if don't do it" (like the DI thing).

**Comment by jtschuster** [comment]
Reopening this issue as it looks like we still are missing some warnings on generic parameters based on tests in src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/GenericParameterWarningLocation.cs
